packages/go_router/lib/src/delegate.dart

```diff
@@ -42,13 +43,52 @@ class GoRouterDelegate extends RouterDelegate<RouteMatchList>
   /// Set to true to disable creating history entries on the web.
   final bool routerNeglect;
 
-  final GlobalKey<NavigatorState> _key = GlobalKey<NavigatorState>();
-
-  RouteMatchList _matches = RouteMatchList.empty();
+  RouteMatchList _matchList = RouteMatchList.empty();
   final Map<String, int> _pushCounts = <String, int>{};
+  final RouteConfiguration _configuration;
+
+  @override
+  Future<bool> popRoute() async {
+    // Iterate backwards through the RouteMatchList until seeing a GoRoute with
+    // a non-null parentNavigatorKey or a ShellRoute with a non-null
+    // parentNavigatorKey and pop from that Navigator instead of the root.
+    final int matchCount = _matchList.matches.length;
+    for (int i = matchCount - 1; i >= 0; i -= 1) {
+      final RouteMatch match = _matchList.matches[i];
+      final RouteBase route = match.route;
+
+      if (route is GoRoute && route.parentNavigatorKey != null) {
+        final bool didPop =
+            await route.parentNavigatorKey!.currentState!.maybePop();
+        // It should not be possible for a GoRoute with parentNavigatorKey to be
+        // the only page, so maybePop should never return false in this case.
+        assert(didPop);
+        return didPop;
+      } else if (route is ShellRoute) {
+        final bool didPop = await route.navigatorKey.currentState!.maybePop();
+        if (didPop) {
+          return didPop;
+        }
```

Comment:
we should continue trying to pop the parentNavigatorKey if this shell route has one

Reply:
Discussed offline, we could add this feature separately if we need to: https://github.com/flutter/flutter/issues/111678